### ISSUE TITLE
add the ability to specify an index file to fai_load

### ISSUE THREAD

**Issue by AndreasHeger** [samtools/htslib#375]
Hi all,

a suggested feature: BAM/CRAM files etc permit the loading of a named index. Could such an argument be added to fai_load?

See here: https://github.com/pysam-developers/pysam/issues/270

Best wishes,
Andreas

**Comment by jeromekelleher** [comment]
This would be very useful for the GA4GH reference server, as we could then have remote files backing references as well as variants and reads. Auto downloading index files does not work well in the context of a server.

**Comment by daviesrob** [comment]
OK, poll here.  It's easy enough to add an interface like:
```
faidx_t *fai_load2(const char *fn, const char *fai_name, const char *gzi_name);
```
but would you want to always have to specify `fai_name` and `gzi_name` (if `fn` is compressed).  Or would it be better to allow either to be passed as NULL and then generate the name by adding the appropriate suffix to `fn`?

The latter would reduce `fai_load` to the trivial function:
```
faidx_t *fai_load(const char *fn) {
    return fai_load2(fn, NULL, NULL);
}
```

I think I prefer the second as it would do the right thing for most people, but I'm open to discussion on the subject.

**Comment by jmarshall** [comment]
Similar question arose for the last batch of these functions in e9b7db4474ee82fc33c883cc906985ec56b32802, and was answered by allowing NULL for the default filenames.  Makes it more flexible to use, e.g., when the extra filename comes from an optional command line option.  See e.g. [bam_index.c](https://github.com/samtools/samtools/blob/develop/bam_index.c#L73), which is less painful than

```c
if (argv[optind+1])
    ret = sam_index_build2(argv[optind], argv[optind+1], csi? min_shift : 0);
else
    ret = sam_index_build(argv[optind], csi? min_shift : 0);
```

Another interesting question is: if you explicitly specify a fai filename and it can't be opened, should that be an error or should it (as now) go ahead and build it for you?

(The ugly convention in these things is to have parameter names like `fnfai` and `fngzi` or so.)

**Comment by jmarshall** [comment]
Also, should this one be `fai_load3()`? :smile:

**Comment by jeromekelleher** [comment]
> Another interesting question is: if you explicitly specify a fai filename and it can't be opened, should that be an error or should it (as now) go ahead and build it for you?

It would be very, very much preferred from our perspective to throw an error back to the user in this situation. As a server running in a daemon process under the www user, trying to write a file  to CWD, or blocking for minutes on a simple file open is never the right thing to do. It would be great if fai_load3 just failed gracefully and returned an informative error code whenever something unexpected happens.

**Comment by jmarshall** [comment]
This response from a server daemon author does not surprise :wink:.  Perhaps we should consider an `int flags` parameter and a `CREATE_MISSING_INDICES` enum for it, so this can be a  choice by the client code.

**Comment by jkbonfield** [comment]
I've just been looking the analogous idx_save functions and they basically go down the route of an extra argument which may be NULL to indicate the default name.  (Rather confusing bouncing back and forth between two functions in the process.)

Hence I think your method for making fai_load a trivial wrapper to fai_load2 to be the more consistent approach.

**Comment by jeromekelleher** [comment]
Another option I guess would be to have integer error codes returned from ``fai_load3()`` which ``fai_load`` can detect and take the appropriate actions. This will make ``fai_load`` more complicated, but would considerably simplify the semantics of ``fai_load3``. To me, library-functions with side effects is an anti-pattern: functionality like automatically building an index if it doesn't exist belongs at the application layer. Of course the semantics of the existing functions like ``fai_load`` must be kept for compatibility, but I would argue that new functions added to the API should strive to be more 'library-like', and just fail with error status when something unexpected happens rather than try to predict what a user might want.

This is just my perspective as a server developer though; I'm sure others might think differently. Thanks for bringing this up @jmarshall!

**Comment by daviesrob** [comment]
htslib's error reporting is a whole other can o'worms :cry:.  I can at least wrap the fprintf's up with
`if (hts_verbose >= 1) { ... }` which would allow you to stop it from scribbling to stderr.

It can only return `NULL` or a pointer, so sending back anything more specific is tricky.  Hopefully `errno` will make sense on most failures.

**Comment by dkj** [comment]
Fixed by 84a89da, see #468
